[MRELEASE-1054] Support for excluding submodules changes.

wuwen5 opened the pull request:
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MRELEASE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MRELEASE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MRELEASE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify -Prun-its` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).


As described in [MRELEASE-1054], when I use the Git submodule in a project, `release: prepare` will fail. I really need to support the release of the Git submodule, and I also have a pull request in the maven-scm project https://github.com/apache/maven-scm/pull/179.

elharo commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractMapVersionsPhase.java:
```diff
@@ -39,6 +43,7 @@
 import org.apache.maven.shared.release.versions.VersionParseException;
 import org.codehaus.plexus.components.interactivity.Prompter;
 import org.codehaus.plexus.components.interactivity.PrompterException;
+import org.codehaus.plexus.util.SelectorUtils;
```
I'd like to avoid addition al plexus dependencies if at all possible.

wuwen5 replied:
I think this should be handled as an independent issue, as it is currently being used in multiple places

elharo replied:
Removing existing usages is a separate issue. Not adding new usages is still important for each PR.

wuwen5 replied:
Removed use of `SelectorUtils`

elharo commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractMapVersionsPhase.java:
```diff
@@ -118,6 +125,15 @@ public ReleaseResult execute(
+        if (additionalExcludes != null) {
+            for (String additionalExclude : additionalExcludes) {
+                exclusionPatterns.add(
+                        additionalExclude.replace("\\", File.separator).replace("/", File.separator));
```
File.separator is platform dependent, but I'm not sure it should be here. That is, exclusions should be specified in a platform independent, standard way. I could be wrong about that but if I am, please elaborate in detail on how this works with references to docs.

wuwen5 replied:
I fully referenced the implementation in `ScmCheckModificationsPhase`.
There is a comment explaining `// SelectorUtils expects File.separator, don't standardize`

elharo replied:
I'm having trouble finding that reference. Regardless, we shouldn't be using plexus for this any more,. Possibly maven-shared-utils but ideally not even that.

elharo commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractMapVersionsPhase.java:
```diff
@@ -165,17 +181,24 @@ public ReleaseResult execute(
+                final String path = root.relativize(pom).getPath();
+
+                if (exclusionPatterns.stream()
+                        .noneMatch(exclusionPattern -> SelectorUtils.matchPath(exclusionPattern, path))) {
```
a basic loop with no lambdas would be easier to read

elharo commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractRewritePomsPhase.java:
```diff
@@ -142,6 +147,15 @@ public ReleaseResult execute(
+        if (additionalExcludes != null) {
+            for (String additionalExclude : additionalExcludes) {
+                exclusionPatterns.add(
+                        additionalExclude.replace("\\", File.separator).replace("/", File.separator));
```
same as above

elharo commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractRewritePomsPhase.java:
```diff
@@ -199,13 +213,18 @@ private void transform(
+
+            if (exclusionPatterns.stream()
+                    .noneMatch(exclusionPattern -> SelectorUtils.matchPath(exclusionPattern, path))) {
+                logInfo(
```
why log this? debug level at most

wuwen5 replied:
This log is not a newly added code, it only adds exclusion conditions here, Please refer to lines `202-206`.

elharo replied:
I wouldn't log any of this at info level, and I certainly add more logging to the existing logging at info level.

wuwen5 commented:
@elharo  I have addressed the issues you pointed out, and I kindly request you to review the modified code. If you have any further suggestions or if there are additional changes needed, please let me know.

elharo commented on maven-release-manager/src/test/java/org/apache/maven/shared/release/phase/MapVersionsPhaseTest.java:
```diff
@@ -2452,6 +2453,9 @@ private static MavenProject createProject(String artifactId, String version) {
         model.setVersion(version);
-        return new MavenProject(model);
+
+        MavenProject mavenProject = new MavenProject(model);
```
does this actually test the bug? It looks like a little new setup code but I don't see how the behavior has changed.

wuwen5 replied:
In the previous test cases, the `file` field in the mock object of `MavenProject` was null, which led to an exception. To address this, null detection has been implemented, and new test cases have been added.

elharo replied:
Thanks. That helps. However, test coverage still feels very limited. There's a lot of new code here and I'm not sure it's all being hit.

wuwen5 replied:
I have added new test cases and I believe they have all been hit

elharo commented on maven-release-manager/src/test/java/org/apache/maven/shared/release/phase/MapVersionsPhaseTest.java:
```diff
@@ -2447,11 +2448,44 @@ public void testUpdateDevelopmentInvalidDefaultDevelopmentVersion_NonInteractive
+        phase.simulate(ReleaseUtils.buildReleaseDescriptor(builder), new DefaultReleaseEnvironment(), reactorProjects);
+
+        // verify
+        assertNull(
```
When would this not be null? Is there a test for that case?

elharo commented on maven-release-manager/src/main/java/org/apache/maven/shared/release/phase/AbstractScmCommitPhase.java:
```diff
@@ -288,11 +303,25 @@ protected static List<File> createPomFiles(ReleaseDescriptor releaseDescriptor,
         for (MavenProject project : reactorProjects) {
-            pomFiles.addAll(createPomFiles(releaseDescriptor, project));
+
+            final String path = project.getFile().getPath();
```
Might want to use a lambda and anyMatch here. Up to you.

wuwen5 replied:
I'll use `anyMatch`
